### ISSUE TITLE
Support usage from Groovy files

### ISSUE THREAD

**Issue by almogtavor** [DanySK/gradle-pre-commit-git-hooks#118]
Is there a possibility of adding support for using the plugin from a Groovy based project? Or any workaround?

**Comment by DanySK** [comment]
Hi @almogtavor.
in principle, the plugin should work regardless of the language being used to write the build file. If someone is willing to try and document how to use the plugin in Groovy, I'll gladly accept a PR that updates the documentation (or the plugin code, if it improves the experience for Groovy users, as far as the functionality when used from Kotlin is unchanged).

**Comment by almogtavor** [comment]
Currently it doesn’t work since groovy won’t recognize the `gitHooks` task. Do you have any idea why this could happen?

**Comment by DanySK** [comment]
Uhmm... the `gitHooks` block is an extension created on the settings file - it should work in both cases

**Comment by DanySK** [comment]
Can you be more specific and link a build scan and the output of your attempts? I need to see the error

**Comment by almogtavor** [comment]
The error I'm getting is 
`> No signature of method: settings_27xxnymre4hp11j1ud1g2gjwl.gitHooks() is applicable for argument types: (settings_27xxnymre4hp11j1ud1g2gjwl$_run_closure1) values: [settings_27xxnymre4hp11j1ud1g2gjwl$_run_closure1@417206fd]`

`Caused by: groovy.lang.MissingMethodException: No signature of method: settings_27xxnymre4hp11j1ud1g2gjwl.gitHooks() is applicable for argument types: (settings_27xxnymre4hp11j1ud1g2gjwl$_run_closure1) values: [settings_27xxnymre4hp11j1ud1g2gjwl$_run_closure1@417206fd]`

My project contains only `build.gradle` and `settings.gradle`. The `build.gradle` is a default `build.gradle` of:
```groovy
plugins {
    id 'java'
}

group 'org.example'
version '1.0-SNAPSHOT'
```

And a `settings.gradle` of:
```groovy
plugins {
    id("org.danilopianini.gradle-pre-commit-git-hooks") version "1.0.12"
}

rootProject.name = 'gradle-pre-commit-git-hooks-groovy-bug'

gitHooks {
    createHooks()
}
```

**Comment by DanySK** [comment]
Oh, I see. Groovy's lambdas are not type-compatible with Kotlin's lambdas.

The Groovy build files *could* be supported, but that requires basically a second DSL. Is it impossible, in your case, to switch the Groovy `settings.gradle` file into a `settings.gradle.kts`, leaving the remainder of the build unchanged? That would allow you to use the plugin without waiting for me preparing a second implementation.

**Comment by almogtavor** [comment]
@DanySK I'll check it. If you'll transform the Kotlin lambda to a regular method do you think it will work?

**Comment by DanySK** [comment]
I would need to tarnish the code a bit, as instead of Kotlin functions with receiver I'll need something that produces bytecode compatible with the Groovy way of reading it.

https://github.com/DanySK/gradle-pre-commit-git-hooks/blob/ad7794a24815b98cea278f38df59d6d37f733253/src/main/kotlin/org/gradle/kotlin/dsl/GradleGitHooksExtensions.kt#L10-L12

I'll probably need to study the Gradle Java API to do so...

**Comment by DanySK** [comment]
It will call the first one, but then I use Kotlin functions with receiver along the script.
Also (and that is a requirement) I do not want to compromise the usage from Kotlin scripts.

**Comment by almogtavor** [comment]
@DanySK So what's the solution you suggest?

**Comment by DanySK** [comment]
I might have found a way to call Kotlin's functions with receiver from Groovy. Can you try:

```groovy
plugins {
    id("org.danilopianini.gradle-pre-commit-git-hooks") version "1.0.14"
}

gitHooks { extension ->
    extension.with {
        commitMsg { context ->
            context.with {
                conventionalCommits { conventionalCommits ->
                    conventionalCommits.with {
                        defaultTypes()
                    }
                }
            }
        }
        createHooks(true)
    }
}
```

The idea is: the JVM signature has the receiver as first argument, so it should be invokable from Groovy by passing a lambda that intercepts the receiver as the first argument.

**Comment by almogtavor** [comment]
@DanySK It works. Thank you very much! In general I think this isn't ideal and pretty hard to understand. Maybe the ideal solution shall be to release a dedicated plugin for Kotlin and dedicated plugin for Groovy?

**Comment by almogtavor** [comment]
@DanySK I totally agree that Groovy sucks and Kotlin DSL is much cleaner and preferable in general. But sometimes you end up with a team that sticks to the Groovy since everything is already written like that. Anyway I'd like to contribute the Java classes that's needed. What do you think should be implemented except from the extension class?

**Comment by DanySK** [comment]
My suggestion is to add several tests using Groovy (covering all your use cases) and then add the required Java methods. They could be implemented by calling Kotlin. I'd be okay with a Groovy implementation, too, if that is easier for you and your team
